src/FwdState.cc

```diff
@@ -801,14 +804,22 @@ FwdState::noteConnection(HappyConnOpener::Answer &answer)
         Must(!Comm::IsConnOpen(answer.conn));
         answer.error.clear(); // preserve error for errorSendComplete()
     } else if (!Comm::IsConnOpen(answer.conn) || fd_table[answer.conn->fd].closing()) {
+        // We do not know exactly why the connection got closed, so we play it
+        // safe: Do not ban retries (i.e. do not set flags.dont_retry) but do
+        // not retry this potentially problematic destination either (i.e. do
+        // not set destinationReceipt).
         syncHierNote(answer.conn, request->url.host());
         closePendingConnection(answer.conn, "conn was closed while waiting for noteConnection");
         error = new ErrorState(ERR_CANNOT_FORWARD, Http::scServiceUnavailable, request, al);
+    } else {
+        assert(!error);
+        destinationReceipt = answer.conn;
```

Comment:
Since a connection can be retried/returned only if answer.reused is true (so that pconnRace becomes racePossible), we probably need to assign destinationReceipt below, in the 'if (answer.reused)' block.

Also, developing this idea further, we could even step back and continue to use the single serverConn of PeerConnectionPointer type (i.e., get rid of destinationReceipt). In this case, syncWithServerConn() below would not change, but the one in FwdState::successfullyConnectedToPeer() would create PeerConnectionPointer in place, passing 'npos' as the position (since that connection is not returnable anyway). Do you think it is worth doing?

Reply:
> a connection can be retried/returned only if answer.reused is true

I doubt it is a good idea to rely on that (currently true but) loosely associated condition.

> we probably need to assign destinationReceipt below, in the 'if (answer.reused)' block.

By "need", you mean "may", right? In other words, the current code does not have a bug, even though it does assign destinationReceipt in some cases where destinationReceipt will not be used, correct?

> Do you think it is worth doing?

I do not think so. I am really worried about making serverConn truthy when there is no usable to-server connection yet. We make complex decisions using serverConn state. I would keep it nil until we have that usable connection.

Reply:
> By "need", you mean "may", right? In other words, the current code does not have a bug

Correct.

Reply:
>  I am really worried about making serverConn truthy when there is no usable to-server connection yet.

Note that I did not suggest to do that. In my suggestion, serverConn would become 'truthy' either immediately in noteConnection() if it is 'reused' (i.e., no further steps are expected) or after all required steps (FwdState::successfullyConnectedToPeer()).

Reply:
Several options would "work" here, but I think we should observe the following design limits:

1. A nil PeerConnectionPointer should not store a destination pos. While a falsy {nil,pos} combination is technically possible, I think it would go against the notion that PeerConnectionPointer is a (smart) pointer. Nil pointers should not store extractable metadata. Nil has/stores "nothing".

2. HappyConnOpener and noteConnection() must not predict whether FwdState will want to call returnPath() on failures. The "which paths should be retried" logic is already quite complex and may become even more tricky in the future. Today, we can predict a negative decision, but, IMO, we should not build a design that precludes future dynamic (i.e. not-predictable) decisions in that area.

With the above boundaries in mind, I do not see how we can avoid storing destinationReceipt.